### DISCUSSION TITLE
typing for a classmethod in a utility class

### DISCUSSION THREAD

**Discussion by jvanasco** [sqlalchemy/sqlalchemy#11254] [Usage Questions] [answered]
I have a utility class `UtilityObject` that is used to offer some basic functionality in many models.  In this example, there is a `get__by__id` function, which uses an `id` column by default – but the class can be figured to use alternate column(s) for the id.

I can't seem to get the typing right for the return value.    Right now it is just returning an `Optional["UtilityObject"]:`.  The two major issues from this in mypy are that:

1)  I should be returning a specific descendant of UtilityObject (somewhat an issue), and
2) the UtilityObject does not have an `id` attribute (this is the bigger issue).  

Can anyone offer better insight on how to handle typing on the example below?

Example Setup

```py
_metadata = MetaData()
Base = declarative_base(metadata=_metadata)


class CoreObject(object):
    __table_pkey__: Optional[str] = None


class UtilityObject(CoreObject):
    @classmethod
    def get__by__id(
        cls,
        dbSession: Session,
        id: Union[str, int],
        id_column: str = "id",
    ) -> Optional["UtilityObject"]:
        if (
            not hasattr(cls, id_column)
            and hasattr(cls, "__table_pkey__")
            and cls.__table_pkey__ is not None
        ):
            id_column = cls.__table_pkey__
        id_col = getattr(cls, id_column)
        if isinstance(id, (list, tuple)):
            return dbSession.query(cls).filter(id_col.in_(id)).all()
        else:
            id_dict = {id_column: id}
            return dbSession.query(cls).filter_by(**id_dict).first()


class FooObject(Base, UtilityObject):
    __tablename__ = "foo_object"
    id: Mapped[int] = mapped_column(Integer, primary_key=True)
```

Example test Usage

```py
class TestModelObjectFunctions(unittest.TestCase):
    def test_one(self):
        dbSession = sess()
        foo = FooObject.get__by__id(dbSession, 1)
        self.assertIsNotNone(foo)
        self.assertEqual(foo.id, 1)
```

**Comment by CaselIT** [comment, marked as answer]
Hi Jonathan,

> 1. I should be returning a specific descendant of UtilityObject (somewhat an issue)

I think you can do `def get__by__id(...) -> Optional[Self]`

> 2. the UtilityObject does not have an id attribute (this is the bigger issue).

Here you probably have to lye to the type checker doing
```py
class UtilityObject(CoreObject):
  if TYPE_CHECKING:
    id: Mapped[int]
  ...
```

You could even do that in a subclass so you can decide depending on the class what of the two is appropriate
```py
class UtilityObjectWithId(UtilityObject):
  if TYPE_CHECKING:
    id: Mapped[int]
```

**Reply by jvanasco** [reply]
Thank you so much, Federico!

The changes needed were only:

```
from typing_extensions import Self
....
def get__by__id(...) -> Optional[Self]
```

This package maintains compat to before `typing.Self` was introduced to the stdlib.  This is also how/why I never knew about `Self`.

You saved the day, again!  I struggled with this for a few hours!

**Reply by CaselIT** [reply]
Do check the additions in typing_extensions. Using only what was available in something like <3.10 is really a pain
